### PR TITLE
[FLINK-30341] Introduce audit_log system table

### PR DESCRIPTION
**Pull request by JingsongLi**
In some scenarios, users need to get the changelog to do some auditing work, such as determining the number of updates and inserts.

We can provide audit_log system table, users can get the rowkind information column.

```
INSERT INTO T VALUES ('1', '2', '3');
INSERT INTO T VALUES ('1', '4', '5');

SELECT * FROM T$audit_log;
```

users can get:
- "+I", "1", "2", "3"
- "-U", "1", "2", "3";
- "+U", "1", "4", "5"

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/AuditLogTable.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            return transform(dataRead.createReader(split), this::convertRow);
+        }
+
+        private RowData convertRow(RowData data) {
+            return new AuditLogRowData(readProjection, data);
+        }
+    }
+
+    /** A {@link MappingRowData} which returns row kind when mapping index is negative. */
+    private static class AuditLogRowData extends MappingRowData {
```
Comment (tsreaper) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/AuditLogTable.java:247]: No need to specially introduce a `MappingRowData`. Just keep the old `ProjectedRowData` and let this class extend `ProjectedRowData`.

File: flink-table-store-core/src/test/java/org/apache/flink/table/store/table/ChangelogWithKeyFileStoreTableTest.java [hunk 2]
```diff
@@ -615,6 +618,69 @@ public void testIncrementalScanOverwrite() throws Exception {
...
+        scan = auditLogTable.newScan().withFilter(rowKindEqual);
+        read = auditLogTable.newRead().withFilter(rowKindEqual);
+        result = getResult(read, scan.plan().splits(), rowDataToString);
+        assertThat(result).containsExactlyInAnyOrder("+I[+I, 2, 20, 200]", "+I[+I, 1, 10, 100]");
+
+        // Read by filter
+        scan = auditLogTable.newScan().withFilter(predicateBuilder.equal(2, 10));
+        read = auditLogTable.newRead().withFilter(predicateBuilder.equal(2, 10));
+        result = getResult(read, scan.plan().splits(), rowDataToString);
+        assertThat(result).containsExactlyInAnyOrder("+I[+I, 1, 10, 100]");
```
Comment (tsreaper) [inline flink-table-store-core/src/test/java/org/apache/flink/table/store/table/ChangelogWithKeyFileStoreTableTest.java:660]: There is only one snapshot and one file, so these tests are actually useless. Create at least 2 files so the predicate actually filters something when scanning files.

Comment (JingsongLi) [reply flink-table-store-core/src/test/java/org/apache/flink/table/store/table/ChangelogWithKeyFileStoreTableTest.java]: Actually, there are two files for two partitions. I'll add a new commit too.

File: flink-table-store-core/src/test/java/org/apache/flink/table/store/table/ChangelogWithKeyFileStoreTableTest.java [hunk 3]
```diff
@@ -615,6 +618,79 @@ public void testIncrementalScanOverwrite() throws Exception {
...
+        TableWrite write = table.newWrite(commitUser);
+        TableCommit commit = table.newCommit(commitUser);
+        write.write(rowDataWithKind(RowKind.INSERT, 1, 10, 100L));
+        write.write(rowDataWithKind(RowKind.INSERT, 2, 20, 200L));
+        commit.commit(0, write.prepareCommit(true, 0));
+
+        write = table.newWrite(commitUser);
+        commit = table.newCommit(commitUser);
+        write.write(rowDataWithKind(RowKind.UPDATE_AFTER, 1, 30, 300L));
+        commit.commit(1, write.prepareCommit(true, 1));
```
Comment (tsreaper) [inline flink-table-store-core/src/test/java/org/apache/flink/table/store/table/ChangelogWithKeyFileStoreTableTest.java:635]: No need to create new write and commit. Also you need to close them before the test ends.

**Review by tsreaper — APPROVED** [review]
Looks good to me. Let's wait for the tests to pass.

**Pull request merged**